cabal-parser/src/Text/Cabal/Value.hs

```diff
@@ -0,0 +1,161 @@
+  pure [Value (T.pack val) (Annotation Nothing loc)]
+ where
+  endChars = ['\r', '\n'] ++ if inBraces then ['{', '}'] else []
+
+{- | Parser to parse module paths.
+
+  The module paths may start with a comma,
+  then an arbitrary number of white spaces
+  then a module path which may be wrapped in apostrophes
+  and consists of words separated by dots,
+  after the module path may be a comma before which there
+  could be an arbitrary number of white spaces.
+-}
+moduleParser :: Bool -> Parser [ValueItem]
+moduleParser inBraces = do
+  vals <- some $ try $ do
+    void $ many " "
+    (val, loc) <- annotateSrcLoc $ do
+      c1 <- option "" $ string ","
+      s1 <- many $ char ' '
+      val' <- between (char '\"') (char '\"') moduleNameParser <|> moduleNameParser
+      s2 <- many $ char ' '
+      c2 <- option "" $ string ","
+      let val = c1 <> T.pack s1 <> T.pack val' <> T.pack s2 <> c2
+      pure val
+    pure $ Value val (Annotation Nothing loc)
+  hspace
+  choice endParsers
+  pure vals
+ where
+  moduleNameParser = some (alphaNumChar <|> char '.' <|> char '_')
+  endParsers = [eof, void eol] ++ ([void $ lookAhead $ string "}" | inBraces])
+
+{- | Parser to parse file paths.
+
+  The file paths may start with a comma,
+  then an arbitrary number of white spaces
+  then a file path which may be wrapped in apostrophes,
+  after the file path may be a comma before which there
+  could be an arbitrary number of white spaces.
+-}
+filepathParser :: Bool -> Parser [ValueItem]
+filepathParser inBraces = do
+  vals <- some $ try $ do
+    void $ many " "
+    (val, loc) <- annotateSrcLoc $ do
+      c1 <- option "" $ string ","
+      s1 <- many $ char ' '
+      val' <- wrappedFpParser <|> filePathParser
+      s2 <- many $ char ' '
+      c2 <- option "" $ string ","
+      let val = c1 <> T.pack s1 <> T.pack val' <> T.pack s2 <> c2
+      pure val
+    pure $ Value val (Annotation Nothing loc)
+  hspace
+  choice endParsers
+  pure vals
+ where
+  wrappedFpParser = between (char '\"') (char '\"') $ many printChar
+  filePathParser = some $ noneOf $ [' ', ',', '\r', '\n'] ++ if inBraces then ['{', '}'] else []
+  endParsers = [eof, void eol] ++ ([void $ lookAhead $ string "}" | inBraces])
+
+parseSourcePosToRange :: SourcePos -> SourcePos -> LSP.Range
+parseSourcePosToRange start end =
+  LSP.Range
+    { LSP._start = parseSourcePosToPosition start
+    , LSP._end = parseSourcePosToPosition end
+    }
+
+parseSourcePosToPosition :: SourcePos -> LSP.Position
+parseSourcePosToPosition (SourcePos _ srcLine srcCol) =
+  LSP.Position
+    { LSP._line = fromIntegral $ unPos srcLine - 1
+    , LSP._character = fromIntegral $ unPos srcCol - 1
+    }
+
+{- | Calls a given parser and returns the value returned by the parser
+  and the range from the start of the parsed value to the end of the parsed value.
+-}
+annotateSrcLoc :: Parser a -> Parser (a, LSP.Range)
+annotateSrcLoc a = do
+  start <- getSourcePos
+  parsed <- a
+  end <- getSourcePos
```

Comment:
Why cabal parser implemented by cabal would not work for this task?

https://github.com/haskell/cabal/tree/master/Cabal-syntax

Reply:
It would work, given the existence https://github.com/Bodigrim/cabal-add/. However, this was deemed a somewhat interesting idea to add a cabal exactprinter, but it is likely not the best way forward.